prover/prover_fri/src/gpu_prover_availability_checker.rs

```diff
@@ -0,0 +1,81 @@
+#[cfg(feature = "gpu")]
+pub mod availability_checker {
+    use std::time::Duration;
+
+    use prover_dal::{ConnectionPool, Prover, ProverDal};
+    use zksync_types::prover_dal::{GpuProverInstanceStatus, SocketAddress};
+
+    use crate::metrics::{KillingReason, METRICS};
+
+    /// Availability checker is a task that periodically checks the status of the prover instance in the database.
+    /// If the prover instance is not found in the database or marked as dead, the availability checker will shut down the prover.
+    pub struct AvailabilityChecker {
+        address: SocketAddress,
+        zone: String,
+        polling_interval: Duration,
+        pool: ConnectionPool<Prover>,
+    }
+
+    impl AvailabilityChecker {
+        pub fn new(
+            address: SocketAddress,
+            zone: String,
+            polling_interval_secs: u32,
+            pool: ConnectionPool<Prover>,
+        ) -> Self {
+            Self {
+                address,
+                zone,
+                polling_interval: Duration::from_secs(polling_interval_secs as u64),
+                pool,
+            }
+        }
+
+        pub async fn run(
+            self,
+            stop_receiver: tokio::sync::watch::Receiver<bool>,
+        ) -> anyhow::Result<()> {
+            while !*stop_receiver.borrow() {
+                let status = self
+                    .pool
+                    .connection()
+                    .await
+                    .unwrap()
+                    .fri_gpu_prover_queue_dal()
+                    .get_prover_instance_status(self.address.clone(), self.zone.clone())
+                    .await;
+
+                // If the prover instance is not found in the database or marked as dead, we should shut down the prover
+                match status {
+                    None => {
+                        METRICS.zombie_prover_instances_count[&KillingReason::Absent].inc();
```

Comment:
what is a possible scenario that would result in this case? I only see bugs leading to this, so this should be a `warn`  or even `error`. Seeing how it is handled here makes me think I'm missing smt

Reply:
It can appear in case the archiver have cleaned up dead instance first and only after that the check was executed, so it will be absent in the db

Reply:
If thinking about it once again, it shouldn't just simply appear - it will depend on the env variables set. If we stick to what we have now(archiving_interval=24h, archive_after=48h, check_interval=30m) - checker will always kill the prover first(because the updated_at value in this prover will be too small), but if someone will change the variables not in the right way - we will be able to still handle this situation and the prover will be killed. I would say that if we will receive Absent status - it is anyway worse than just dead, so I agree that it would be good to have warn or error, probably the alert for such status should have higher priority as well

Reply:
>  I'd defer to @RomanBrodetski if he wants more specific level.

I don't want any particular level, my question is more "if you want to notice that such events started to happen". E.g.  if it's indicattive of bug, you probably want to learn it sooner. Second Artem's message may suggest we do.

In the end 100% your decision - feel free to keep as is